src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -114,33 +113,19 @@ public void clusterChanged(ClusterChangedEvent event) {
                 }
             }
 
-            // check and write changes in indices
-            for (IndexMetaData indexMetaData : newMetaData) {
-                String writeReason = null;
-                IndexMetaData currentIndexMetaData;
-                if (currentMetaData == null) {
-                    // a new event..., check from the state stored
-                    try {
-                        currentIndexMetaData = metaStateService.loadIndexState(indexMetaData.index());
-                    } catch (IOException ex) {
-                        throw new ElasticsearchException("failed to load index state", ex);
-                    }
-                } else {
-                    currentIndexMetaData = currentMetaData.index(indexMetaData.index());
-                }
-                if (currentIndexMetaData == null) {
-                    writeReason = "freshly created";
-                } else if (currentIndexMetaData.version() != indexMetaData.version()) {
-                    writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
-                }
-
-                // we update the writeReason only if we really need to write it
-                if (writeReason == null) {
-                    continue;
-                }
+            Iterable<IndexMetaWriteInfo> writeInfo;
+            if (isDataOnlyNode(event.state())) {
+                writeInfo = filterStateOnDataNode(event, currentMetaData);
```

Comment:
I know this was already there, but can we rename currentMetaData to previouslyWrittenMetaData ? I think it will be easy to understand what's what here..